ssl/quic/quic_port.c

```diff
@@ -771,40 +752,30 @@ static int marshal_validation_token(QUIC_VALIDATION_TOKEN *token,
 static int parse_validation_token(QUIC_VALIDATION_TOKEN *token,
                                   const unsigned char *buf, size_t buf_len)
 {
-    PACKET pkt, subpkt;
+    PACKET pkt;
 
     if (buf == NULL || token == NULL)
         return 0;
 
     token->remote_addr = NULL;
 
     if (!PACKET_buf_init(&pkt, buf, buf_len)
-        || !PACKET_get_sub_packet(&pkt, &subpkt, sizeof(token->is_retry))
-        || !PACKET_copy_bytes(&subpkt, &token->is_retry, sizeof(token->is_retry))
+        || !PACKET_copy_bytes(&pkt, &token->is_retry, sizeof(token->is_retry))
         || !(token->is_retry == 0 || token->is_retry == 1)
-        || !PACKET_get_sub_packet(&pkt, &subpkt, sizeof(token->timestamp))
-        || !PACKET_copy_bytes(&subpkt, (unsigned char *)&token->timestamp,
+        || !PACKET_copy_bytes(&pkt, (unsigned char *)&token->timestamp,
                               sizeof(token->timestamp))
-        || !PACKET_get_sub_packet(&pkt, &subpkt, sizeof(token->odcid.id_len))
-        || !PACKET_copy_bytes(&subpkt, &token->odcid.id_len,
-                              sizeof(token->odcid.id_len))
-        || token->odcid.id_len > QUIC_MAX_CONN_ID_LEN
-        || !PACKET_get_sub_packet(&pkt, &subpkt, token->odcid.id_len)
-        || !PACKET_copy_bytes(&subpkt, (unsigned char *)&token->odcid.id,
-                              token->odcid.id_len)
-        || !PACKET_get_sub_packet(&pkt, &subpkt, sizeof(token->rscid.id_len))
-        || !PACKET_copy_bytes(&subpkt, &token->rscid.id_len,
-                              sizeof(token->rscid.id_len))
-        || token->odcid.id_len > QUIC_MAX_CONN_ID_LEN
-        || !PACKET_get_sub_packet(&pkt, &subpkt, token->rscid.id_len)
-        || !PACKET_copy_bytes(&subpkt, (unsigned char *)&token->rscid.id,
-                              token->rscid.id_len)
-        || !PACKET_get_sub_packet(&pkt, &subpkt, sizeof(token->remote_addr_len))
-        || !PACKET_copy_bytes(&subpkt, (unsigned char *)&token->remote_addr_len,
-                              sizeof(token->remote_addr_len))
+        || (token->is_retry
+            && (!PACKET_get_1_len(&pkt, (size_t *)&token->odcid.id_len)
+                || token->odcid.id_len > QUIC_MAX_CONN_ID_LEN
+                || !PACKET_copy_bytes(&pkt, (unsigned char *)&token->odcid.id,
+                                      token->odcid.id_len)
+                || !PACKET_get_1_len(&pkt, (size_t *)&token->rscid.id_len)
```

Comment:
Ah I did not notice you cannot use PACKET_get_1_len() here because id_len is just a byte. Use `PACKET_get_bytes(&pkt, &token->odcid.id_len, sizeof(token->odcid.id_len))` and similarly below.